fumosay.1
.TH fumosay 1 "2025 Apr 06" "version 1.2.3"
.SH NAME
fumosay - funky talking fumofumos

.SH SYNOPSIS
fumosay
.BI "[-hvl] [ngbrR] [-c" "color" "] [-W " "column" "] [-f " "name(s)" "] [-E " "expression" "]"

.SH EXAMPLES
.IP "fumosay Hello world!"
.IP "fortune | fumosay -c -f alice,marisa -E 3"
Have Alice or Marisa read out the output of fortune, with colored text and a smile.

.SH DESCRIPTION
.I fumosay
draws an ASCII art of a FumoFumo plushie saying something the user provided. If no message is in the command line, fumos will read from standard input. The fumo will word-wrap the message at 80 columns and print it out on standard output.

.SH OPTIONS
.TP
.B -l
List the names of all available fumos.
.TP
.B -n
Disable word-wrapping, preserving white-spaces and line breaks. Allowing the fumo to say another fumo.
.TP
.B -g
Display a simple prompt at the top of the message, like a video game dialogue.
.TP
.B -b
Makes the output bold font.
.TP
.B -r/-R
Makes the text rainbow. -R makes the fumo rainbow as well.
.TP
.BI "-c" "R,G,B"
Color the text with a color that vaguely represents the fumo. Optionally, specify an RGB color of your own, there are no spaces after the character c.
.TP
.BR "-W " "column"
Specify where the message should be word-wrapped, the default is 80. This option is
.I ignored
.RB "if " "-n " "is set."
.TP
.BR "-f " "name"
Specify specific fumos to read your message. For more than one name, separate them with ",", one will be picked at random among them.
.TP
.BR "-E " "expression"
Set an alternate expression (mouth string) for the fumo. You may choose a preset from the following or specify your own custom expression.
.RS
.br
\'1\' - frowny
.br
\'2\' - neutral
.br
\'3\' - smile
.br
\'4\' - bigger smile
.br
\'v\' - little vampire sister
.br
\'w\' - lucky bunny :3
.br
\'b\' - is for baka
.br
\'d\' - :D
.br
\'r\' - pick randomly
.PP
To specify a custom expression, enter anything longer than one character. The first 24 characters of the input will be used, if fewer than 24 characters is given, the space will be filled by spaces automatically. The 24 characters will be split into 3 separate lines of 8 characters each. The center of each line is located at the 5th position out of 8. The following example produces a 3-wide smile in the middle of line 2. There are 11 leading spaces.
.sp
fumosay -E "           *_*"
.RE
.TP
.B -h
Display help message.
.TP
.B -v
Display version.

.SH ADDING FUMOS
The quote-unquote configuration file "fumo.fumo" is really just C code with instructions. You can add new fumos to it and recompile fumosay.

.SH BUGS
I don't know of anything significant for now.

.SH SEE ALSO
.IR "cowsay" "(1), " "FORTUNE" "(6)"
.sp
The FumoFumos are adorable friendly plushies (mainly) featuring characters from
.RI "Team Shanghai Alice's " "Touhou Project" ", made by Gift!"
.br
The ASCII arts are created with help from the tool of asciiart.eu.
.br
fumosay is inspired by Tony Monroe's cowsay.


changelog.md
### v1.2.3 - 2025/04/06
Say YA!!!
- Added 4 new fumos:<br>
  The fairies (and spring) have arrived! Sunny, Luna, Star, and Lily are here.
- Added 1 new expression preset:<br>
  Luna does the :D face.
- Newlines in command line messages no longer cut off everything after.
- Now uses getrandom() for all builds.

### v1.2.2 - 2024/11/23
My Youtube Music Recap is 46% Touhou.
- Added 4 new fumos:<br>
  Yuuma, Doremy, Seiga, and Kosuzu!
- Fixed `fsgq2.py` after I forgot to revert the debugging changes again

### v1.2.1 - 2024/10/15
MORE FUMOS
- Added 4 new fumos:<br>
  Byakuren, Medicine, Shinmyoumaru, and Ellen!<br>
  Byakuren even has her own special rainbow colors.
- The 'Reimu says:' text (with `-g`) will now be colored when used with rainbows
- Fixed wrong fumo when the message contains the word `to` in `fsgq2.py`

### v1.2 - 2024/08/09
In remembrance of Hieda no Akyuu.<br>
This is *the* big update until I figure out how to make animated fumos!
- Added Unicode support:<br>
  - Provided by GNU `libunistring`
  - UTF-8 is assumed
  - Everything should display properly now.
- Rainbows & colors work with unicode too!<br>
  - ANSI escape sequences no longer breaks offset
  - You can finally pipe colored fumosay output into fumosay
![Fumoception](https://github.com/randomtwdude/fumosay/blob/main/images/fumoception.png?raw=true)
- Added 4 new fumos:<br>
  - Mystia's Izakaya at the Human Village: Mystia, Kyouko, Akyuu, and Keine!
- Fixed broken output when reading from the shell in Windows (CR strikes again)
- Messages from the command line will now be properly split by space
- No more (unintentional) memory leaks
- Reorganized code files:<br>
  - `fumosay.c` can now compile with the Cosmopolitan toolchain
  - no longer including C files in other C files
- `fsgq2.py` replaces `fsgq.sh`

### v1.1.10 - 2024/07/21
Finally, fumo completeness.
- Added 4 new fumos:<br>
  Orin, Parsee, Nue, and Chimata!
- The lunar rainbow is now available with `-r`:<br>
  - Always active for Chimata
  - `-R` to also make the fumo rainbow
  - Now requires the math library when compiling
- Bold font is now available with `-b`
- The `-f` option is now a real fuzzy search. (Bitap wow!)
  - fumo data now supports multiple names
- Various fumo touch-ups:<br>
  all fumo now have odd number gap in the center (Reimu, Flan, and Tewi were the exceptions)
  Sanae and others gets more hair
- Number of memory leaks reduced by 1 (one), not that it matters anyways.

### v1.1.9 - 2024/07/13
- Added 4 new fumos:<br>
  Seija, Sagume, Nazrin, and Okina!
- Hina has a more proper bow now.
- Added all fumos to the repo.

### v1.1.8 - 2024/06/15
Almost there!
- Added 4 new fumos:<br>
  Kogasa, Okuu, Rinnosuke, and Hecatia!
- Tweaks for Youmu and Keiki
- Technical stuff:<br>
  code for the Cosmopolitan build is now included in the tar package.<br>
  Prebuilt binary now works on Android.

### v1.1.7 - 2024/05/20
#### 05/24 - fixed fumo_picker() not accepting extra words
Ghost of Tsushima may affect the release schedule.
- Added 3 new fumos!<br>
  Shiki Eiki, Keiki, and Hina joins the roster!
- Added `-v`: display version.

### v1.1.6 - 2024/05/14
Help me, Eirin!
- Added 3 new fumos!<br>
  Welcome Renko, Maribel, and Junko to the roster!
- Touched up some other fumos.<br>
  Kaguya, Shion, Eirin, and Miko looking nicer.

fsgq:
- Fixed fumosay installation (hopefully)
- Fixed Gensoquote error message showing

### v1.1.5 - 2024/04/29
History.
- Added 3 new fumos!<br>
  Welcome, Kasen, Miko, and Futo!
- No longer uses Autotools.

fsgq:
- Added to the source tree.
- Handle characters without quotes.

### v1.1.4 - 2024/04/11
It's the Youkai Mountain special.
- Welcome 3 new fumos!<br>
  We have Momiji, Hatate, and Nitori.
- Tabs now behave properly (they align)<br>
  This works in messages from command line arguments, too. (unlike newlines, which continue to be broken)

### v1.1.3 - 2024/03/23
This update is much delayed because it kept segfaulting. Unsigned integers and arithmetics DO NOT mix.
- Brought 4 new fumos. (32/51)<br>
  Friends from the Moon - Reisen (the Gensokyo one) and Eirin.
  Accesories included - Yuuka with a flower and Kokoro with her mask.
- A little bit cleaner code.
- Windows support is now provided by [Cosmopolitan libc](https://github.com/jart/cosmopolitan)<br>
  This one file should run on Linux, Windows, Mac OS, and (Free/Open/Net)BSD, really cool.<br>
  ~~Bonus: won't be misidentified as a virus.~~

### v1.1.2 - 2024/03/13
- Added 3 new fumos. (28/51)<br>
  Welcome Chen, Ran, and Rumia!
- Switched the random source to `arc4random`.<br>
  No longer produces the same result when run twice in one second.
- Minor tweaks for Reimu.
- FIXED: specifying `-f` with unrecognised names causing a crash due to division by zero.
- FIXED: specifying `-c` with less than 3 numbers causing out-of-bound memory access.
- FIXED: accidentally pressed TAB on Flandre while editing...
- Flandre Scarlet now goes by her full name.
- Changed default color of Yukari to a lighter purple.
- Introducing `fsgq`, the fumosay-gensoquote wrapper script.

### v1.1.1 - 2024/03/06
With this update we begin the quest to add every Touhou fumo... with Suika and Shion!

### v1.1 - 2024/02/26
Just a minor update because no classes.
- The `-c` option now accepts optional custom colors. For example, `-c200,150,230` gives a nice purple.
- New fumofumos!<br>
  Welcome Suwako and Meiling!
- Mini tweaks.

### v1.0 - 2024/01/07
The fumos thank you for having them with you! The fumos have elected to bump the version number to 1.
> More fumos to come still!
- New word-wrapping algorithm should produce more "rectangular" and better-looking paragraphs.<br>
  Also fixes a bug where the text box does not render correctly if the last word in the message is longer than line width.
- The `-f` option now accepts names of multiple fumos, separated by `,`. One of them will be chosen at random.
- Added two more expression presets.<br>
  | preset | description |

fumosay.c
// -- fumosay --
// like cowsay, but with funky fumos!
// ᗜ_ᗜ have a nice day ᗜˬᗜ

/* ===== INCLUDES ===== */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <string.h>
#include <stdbool.h>

// Everything fumo
#include "fumoutil.h"
#include "fumo.fumo"
// Everything text
#include "fumolang.h"

#define SET_COLOR(r,g,b) printf("\033[38;2;%hd;%hd;%hdm", r, g, b);
#define RESET_COLOR printf("\033[0m");

#ifndef __COSMOCC__
#define VERSION_STRING "fumosay 1.2.3\n"
#else
#define VERSION_STRING "fumosay 1.2.3 cosmos\n"
#endif

int MAX_WIDTH = 80;
bool isByakuren = false;

/* ===== FUNCTIONS ===== */
void fumo_help(fumo_who fm) {
  printf("fumosay: funky talking fumofumos\n"
         "--------------------------------\n"
         "Usage: fumosay [-hvl] [ngrRb] [-c] [-W column] [-f name] [-E expression] (message)\n"
         "-l     List all fumos.\n"
         "-h     Display this message.\n"
         "-v     Show version.\n"
         "\n"
         "-n     Disable word-wrapping. Overrides -W.\n"
         "-W     Specify roughly where the message should be wrapped.\n"
         "       The default is 80.\n"
         "-f     Pick fumos by names.\n"
         "-E     Change the expression (mouth string) of the fumo.\n"
         "       Use a one-letter preset below or enter your own!\n"
         "       Presets: '1', '2', '3', '4', 'v', 'w', 'b', 'r'(random)\n"
         "-g     Display the character's name above the text box like a story game.\n"
         "-c     Color message. Optionally, specify RGB like \"-c255,255,255\".\n"
         "-r     Rainbow message courtesy of Tenkyuu Chimata.\n"
         "-R     Rainbow fumo as well.\n"
         "-b     Bold font.\n"
         "--------------------------------\n"
         "In spirit of the original cowsay, made by Tony Monroe in 1999.\n"
         "Fumos feature characters from Touhou Project, say hi to ");

  SET_COLOR(FUMO_LIST[fm]->color.R, FUMO_LIST[fm]->color.G, FUMO_LIST[fm]->color.B);
  fputs(FUMO_LIST[fm]->name[0], stdout);
  RESET_COLOR;
  fputs("!\n", stdout);
}

/* Turns a string to number (basically just strtol that rejects 0) */
int numberize(char *c) {
  long number = strtol(c, 0, 0);
  if (number == 0) {
    fprintf(stderr, "ᗜ_ᗜ: Invalid parameter %s\n", c);
  }
  return (int)number;
}

int main(int argc, char **argv) {
  // init
  bool no_wrap = false;
  bool display_name = false;
  char colour = 0;
  color custom_colour = {0, 0, 0};
  bool rainbow = false;
  bool rainbow_fumo = false;
  bool bold = false;
  char expr = 0;
  char custom_expr[30] = {0};
  fumo_who fm = -1;

  // argument
  char opt;
  while ((opt = getopt(argc, argv, "hvlngrRbc::E:W:f:")) != -1) {
    switch (opt) {
    case 'h':;
      fumo_who helper_fumo = random_uniform(FUMO_COUNT);
      fumo_help(helper_fumo);
      fumo_expr(helper_fumo, expr, custom_expr);
      fumo_fumo(helper_fumo, fputs);
      return 0;
    case 'l':
      fumo_list();
      return 0;
    case 'n':
      MAX_WIDTH = 100000000;
      no_wrap = true;
      break;
    case 'W':
      MAX_WIDTH = no_wrap ? MAX_WIDTH : numberize(optarg);
      break;
    case 'f':;
      char *token = strtok(optarg, ",.;/|");
      short choice_count = 0;
      short *choices = NULL;
      while (token) {
        int choice = fumo_picker(token);
        if (choice > -1) {
          choices = realloc(choices, ++choice_count * sizeof(short));
          choices[choice_count - 1] = choice;
        }
        token = strtok(NULL, ",.;/|");
      }
      if (choices) {
        fm = choices[random_uniform(choice_count)];
        free(choices);
      }
      break;
    case 'g':
      display_name = true;
      break;
    case 'c':
      colour = 1;
      if (optarg) {
        colour = 2;
        // custom color values
        char *token = strtok(optarg, ",.;/|");
        custom_colour.R = strtol(token, NULL, 0);
        if ((token = strtok(NULL, ",.;/|"))) {
          custom_colour.G = strtol(token, NULL, 0);
        }
        if ((token = strtok(NULL, ",.;/|"))) {
          custom_colour.B = strtol(token, NULL, 0);
        }
      }
      break;
    case 'r':;
      rainbow = true;
      break;
    case 'R':;
      rainbow = true;
      rainbow_fumo = true;
      break;
    case 'b':;
      bold = true;
      break;
    case 'E':;
      int optlen = strlen(optarg);
      if (optlen > 1) {
        // use custom expression
        short padding = 24 - strlen(optarg);
        if (padding < 0) {
          padding = optarg[24] = 0; // ensure non-negative padding & truncate string
        }
        sprintf(custom_expr, "%s%.*s", optarg,
                padding, "                              ");
      } else {
        expr = optarg[0];
      }
      break;
    case 'v':;
      printf(VERSION_STRING);
      return 0;
    } // switch
  }

  if (MAX_WIDTH < 3) {
    MAX_WIDTH = 3;
  }

  if (bold) {
    fputs("\033[1m", stdout);
  }

  // fumo reads
  int word_count = 0;
  char **word_vector = NULL;
  if (optind == argc) { // read from stdin
    while (1) {
      char *token;
      token = getInput(stdin, 64);
      if (!token) {
        fumo_panic("ᗜxᗜ: fumo panic - memory allocation failed: getInput.\n", expr, custom_expr);
      }
      if (0 == strlen(token)) {
        free(token); /* sneaky 1-byte memory leak exterminated */
        break;
      }
      if (no_wrap) {
        token = replaceTab(token, 8);
        if (!token) {
          fumo_panic("ᗜxᗜ: fumo panic - memory allocation failed: replaceTab.\n", expr, custom_expr);
        }
      }
      word_vector = realloc(word_vector, (word_count + 1) * sizeof(char *));
      if (!word_vector) {
        fumo_panic("ᗜxᗜ: fumo panic - memory allocation failed while reading!\n", expr, custom_expr);
      }
      word_vector[word_count++] = token;
    }

  } else { // read from cmd line arguments
    word_count = argc - optind;

    // I can't realloc argv so this'll have to do
    word_vector = malloc(word_count * sizeof(char *));
    if (!word_vector) {
      fumo_panic("ᗜxᗜ: fumo panic - memory allocation failed while reading!\n", expr, custom_expr);
    }
    memcpy(word_vector, argv + optind, word_count * sizeof(char *));

    char *newline = NULL;
    for (int i = 0; i < word_count; i++) {
      // remove newlines from cmd message
      while ((newline = strchr(word_vector[i], '\n'))) {
        *newline = ' ';
      }
      // replace tabs with spaces
      if (strchr(word_vector[i], '\t') == NULL) {
        continue;
      }
      char *token = strdup(word_vector[i]);
      word_vector[i] = replaceTab(token, 8);
    }
  }

  // cut lines into words
  if (!no_wrap) {
    word_vector = splitWords(word_vector, &word_count);
  }

  // calculate the width of the bubble
  int bubble_width = longestLineWidth(word_count, word_vector) + 1;

  // choose a fumo if not already chosen
  if (fm == -1) {
    fm = random_uniform(FUMO_COUNT);
  }

  // finalize expression
  fumo_expr(fm, expr, custom_expr);

  // colours
  if (fm == 55) {
    rainbow = true;
  } else if (fm == 60) {
    isByakuren = true;
  }

  int (*fumo_say)(const char *, FILE *);
  if (rainbow) {
    fumo_say = &lolfumo;
  } else {
    fumo_say = &fputs;
  }

  if (colour == 1) {
    SET_COLOR(FUMO_LIST[fm]->color.R, FUMO_LIST[fm]->color.G, FUMO_LIST[fm]->color.B);
  } else if (colour == 2) {
    SET_COLOR(custom_colour.R, custom_colour.G, custom_colour.B);
  }

  // CYOA mode
  if (display_name) {
    char cyoa_str[64];
    sprintf(cyoa_str, "%s says:\n", FUMO_LIST[fm]->name[0]);
    fumo_say(cyoa_str, stdout);
  }

  // top line
  fumo_say(" ", stdout);
  for (int i = 0; i < bubble_width; i++) {
    fumo_say("_", stdout);
  }
  fumo_say("\n", stdout);

  // message
  int last_section = 0, cur_section = 0;
  char **process = word_vector;
  char *nl;
  for (int i = 0; i < word_count; i++) {
    cur_section++;
    if ((nl = strchr(word_vector[i], '\n')) != NULL || i == word_count - 1) {
      if (nl) {
        *nl = 0; // strip the last newline of the paragraph
      }
      process += last_section;
      wordWrapper(
        cur_section,
        process,
        bubble_width - 2,
        bubble_width,
        no_wrap,
        optind != argc,
        fumo_say
      );
      last_section = cur_section;
      cur_section = 0;
    }
  }

  // bottom line
  fumo_say(" ", stdout);
  for (int i = 0; i < bubble_width; i++) {
    fumo_say("-", stdout);
  }

  // reset color
  if (colour || (rainbow && !rainbow_fumo)) {
    RESET_COLOR;
    if (bold) {
      fputs("\033[1m", stdout);
    }
  }

  // fumo!
  if (rainbow_fumo) {
    fumo_fumo(fm, &lolfumo);
  } else {
    fumo_fumo(fm, &fputs);
  }

  if (rainbow_fumo || bold) {
    RESET_COLOR;
  }

  free(word_vector);
  return 0;
}
